scd/token.c

```diff
@@ -0,0 +1,321 @@
+		new_token->token_data->type = NONE;
+		break;
+	}
+	case (SOFT): {
+		DEBUG("Create scd_token with internal type 'SOFT'");
+
+		ASSERT(constr_data->uuid);
+		ASSERT(constr_data->str.softtoken_dir);
+
+		token_file = mem_printf("%s/%s%s", constr_data->str.softtoken_dir,
+					constr_data->uuid, STOKEN_DEFAULT_EXT);
+		if (!file_exists(token_file)) {
+			if (softtoken_create_p12(token_file, STOKEN_DEFAULT_PASS,
+						 constr_data->uuid) != 0) {
+				ERROR("could not create new softtoken file");
+				mem_free(token_file);
+				goto err_tk_new;
+			}
+		}
+		new_token->token_data->int_token.softtoken = softtoken_new_from_p12(token_file);
+		if (!new_token->token_data->int_token.softtoken) {
+			ERROR("Creation of softtoken failed");
+			mem_free(token_file);
+			goto err_tk_new;
+		}
+		mem_free(token_file);
+
+		new_token->token_data->type = SOFT;
+		new_token->lock = int_lock_st;
+		new_token->unlock = int_unlock_st;
+		new_token->is_locked = int_is_locked_st;
+		new_token->is_locked_till_reboot = int_is_locked_till_reboot_st;
+		new_token->wrap_key = int_wrap_st;
+		new_token->unwrap_key = int_unwrap_st;
+		new_token->change_passphrase = int_change_pw_st;
+		break;
+	}
+#ifdef ENABLESCHSM
+	case (USB): {
+		DEBUG("Create scd_token with internal type 'USB'");
+
+		ASSERT(constr_data->uuid);
+		ASSERT(constr_data->str.usbtoken_serial);
+
+		new_token->token_data->int_token.usbtoken =
+			usbtoken_init(constr_data->str.usbtoken_serial);
+		if (NULL == new_token->token_data->int_token.usbtoken) {
+			ERROR("Creation of usbtoken failed");
+			goto err_tk_new;
+		}
+		new_token->token_data->type = USB;
+		new_token->lock = int_lock_usb;
+		new_token->unlock = int_unlock_usb;
+		new_token->is_locked = int_is_locked_usb;
+		new_token->is_locked_till_reboot = int_is_locked_till_reboot_usb;
+		new_token->wrap_key = int_wrap_usb;
+		new_token->unwrap_key = int_unwrap_usb;
+		new_token->change_passphrase = int_change_pw_usb;
+		break;
+	}
+#endif // ENABLESCHSM
+	default: {
+		ERROR("Unrecognized token type");
+		mem_free(new_token);
+		return NULL;
+	}
+	}
+
+	return new_token;
+
+err_tk_new:
+	uuid_free(new_token->token_data->token_uuid);
+err_token_uuid:
+	mem_free(new_token->token_data);
+err_token_data:
+	mem_free(new_token);
+
+	return NULL;
+}
+
+scd_tokentype_t
+token_get_type(scd_token_t *token)
+{
+	TRACE("SCD TOKEN: scd_token_get_type");
+
+	ASSERT(token->token_data->type);
+	return token->token_data->type;
+}
+
+uuid_t *
+token_get_uuid(scd_token_t *token)
+{
+	TRACE("SCD TOKEN: scd_token_get_uuid");
+
+	ASSERT(token->token_data->token_uuid);
+	return token->token_data->token_uuid;
+}
+
+static void
+token_data_free(scd_token_data_t *token_data)
+{
+	switch (token_data->type) {
+	case (NONE):
+		break;
+	case (SOFT):
+		TRACE("Removing softtoken %s", uuid_string(token_data->token_uuid));
+		softtoken_remove_p12(token_data->int_token.softtoken);
+		softtoken_free(token_data->int_token.softtoken);
```

Comment:
softtoken_free should remove p12 file

Reply:
no, i don't think so. there is a semantic difference between freeing the scd object "softtoken" and removing the persistent file. the distinction enables to unload token e.g. to preserve memory if a container is stopped but preserve the pkcs12 file fr future use